nixos/modules/services/wayland/hypridle.nix

```diff
@@ -10,17 +10,31 @@ let
 in
 {
   options.services.hypridle = {
-    enable = lib.mkEnableOption "hypridle, Hyprland's idle daemon";
+    enable = lib.mkEnableOption null // {
+      description = ''
+        Whether to enable Hypridle, Hyprland's idle daemon.
+
+        ::: {.note}
+        For this service to work properly on Hyprland, you
+        need to have `programs.hyprland.withUWSM` enabled.
+        :::
+      '';
+    };
     package = lib.mkPackageOption pkgs "hypridle" { };
   };
 
   config = lib.mkIf cfg.enable {
     environment.systemPackages = [ cfg.package ];
 
-    systemd = {
-      packages = [ cfg.package ];
-      user.services.hypridle.wantedBy = [ "graphical-session.target" ];
-      user.services.hypridle.path = [
+    systemd.packages = [ cfg.package ];
+
+    systemd.user.services.hypridle = {
+      # Service should be only be started on Hyprland
+      # this target is started by UWSM
+      wantedBy = [ "wayland-session@Hyprland.target" ];
```

Comment:
```suggestion
      # start only with Hyprland units
      wantedBy = [
        "wayland-session@Hyprland.target"
        "wayland-session@hyprland.desktop.target"
        "wayland-session@hyprland\x2duwsm.target"
      ];
      # order start after compositor (and WAYLAND_DISPLAY) is available
      # order stop before compositor
      after = [ "graphical-session.target" ];
      # stop with the session
      partOf = [ "graphical-session.target" ];
```

Reply:
These two targets do not exist on NixOS. We've explicitly limited this module to only work with the `programs.hyprland.withUWSM` option which only creates the `wayland-session@Hyprland.target`. (I'd prefer a stronger API than the Hyprland package's binary name here but that's something for our uwsm module to offer.

We also want to order hypridle `before` the wayland session so that it's only considered up when hypridle has started up.

Reply:
edited

Reply:
In my config I made it `before` and `partOf` `wayland-session@Hyprland.target` because I think of hypridle as belonging to the "Hyprland session" rather than the graphical-session directly. This would also be okay IMV.

I do find it a bit weird to depend on the session target rather than the wm service though as that's our actual dependency.

Reply:
Units of service and target are bound, so there is no difference for dependencies. Targets are more frequently used as dependencies.
```
$ grep -r ^WantedBy= /lib/systemd/system | grep -cF .target
124
$ grep -r ^WantedBy= /lib/systemd/system | grep -cF .service
90
$ grep -r ^WantedBy= /lib/systemd/user | grep -cF .target
24
$ grep -r ^WantedBy= /lib/systemd/user | grep -cF .service
17
```

Reply:
`wayland-wm@.service` is bound to `wayland-session@.target`, so they are interchangeable as dependencies (but can differ when ordering).

Targets are more frequently used for dependencies than services.

Reply:
Depending on `graphical-session.target`, is fine in my opinion.

When a WM managed by UWSM is terminated, graphical-session also stops. So I understand the partOf bit.

What I don't understand is why are we supposed to start it "before" `graphical-session`? Starting it after isn't correct?

> I'd prefer a stronger API than the Hyprland package's binary name here

Agreed.